### ISSUE TITLE
When to CreateOffer()

### ISSUE THREAD

**Issue by AtosNicoS** [microsoft/MixedReality-WebRTC#344]
**Describe the problem**
When looking at the UWP example code of 1.0.3 I found it confusing when the offer is sent:
https://github.com/microsoft/MixedReality-WebRTC/blob/release/1.0/examples/TestAppUwp/MainPage.xaml.cs#L1223

I have never tried out that example, I have just adopted the mechanisms from it in my code. I am wondering why `CreateOffer()` in the mentioned line is only send when `IsConnected`. In my code the connection will only be established, when I create an offer. In which scenario is the mentioned `CreateOffer` ever executed? Am I missing something? How is an offer then sent? Only with the manual button?

It would be nice to have some additional information about those properties, preferred in the Docs. Also about the renegotiate mechanism. What it is used for, the current limitations (see the TODOs in the example) and how to implement it properly.

BTW: There are a lot TODOs in the official example. This makes it even harder for me to use the library, as it seems broken/unfinished on some parts?

**Comment by djee-ms** [comment]
>  I am wondering why CreateOffer() in the mentioned line is only send when IsConnected.

Because if not, when the user starts the local webcam playback, this will immediately try to negotiate with the remote peer, which means that if the peer is not connected it will "make a new call" (first negotiation), which is not what a user starting its local webcam expects (Skype doesn't call your friends when you test your webcam locally, only when you actually press the Call button).

> In which scenario is the mentioned CreateOffer ever executed?

When the tracks change while a call is on-going. The "call" concept is a high-level conceptual notion, which helps understand the whole process, and is often a concept in the application itself. But at the WebRTC level, there's no difference between "making a call" (first negotiation) and "adding/removing a track from an on-going call" (renegotiation); they both require a call to `CreateOffer()` to (re-)negotiate a track setup.

> How is an offer then sent? Only with the manual button?

On very first negotation, yes, only via the manual button, so that we don't start a call without the user's consent.

After that, either when we know the tracks changed since the local user made the change (`StartLocalMediaClicked()` and `StopLocalVideo()`), or when the implementation tells use that they changed (`OnPeerRenegotiationNeeded()`). The reason we don't rely exclusively in this case on `OnPeerRenegotiationNeeded()` is to batch changes: if you add an audio track and a video track, you get a renegotiation needed event between the 2 calls (right after the audio one, before even adding the video one), which means:
1. at best you will negotiate twice, once to add the audio, and a second time to add the video on top of that, although we know about the second change before starting the first one, so that's a large waste of time (2 full round-trips and negotiations to remote peer instead of just 1).
2. at worst, same, but the second offer will start while the first negotiation is not finished, which is an invalid state for a negotiation to start, and that will simply fail the second negotiation (the offer will not be created) and video will not work.

> It would be nice to have some additional information about those properties, preferred in the Docs. Also about the renegotiate mechanism.

Agreed. But we lack a bit the time to do that, and those are generic WebRTC concepts, nothing specific to this project, so we're relying on the vast amount of WebRTC resources out there for users to educate themselves, and focusing on the things specific to this project.

> BTW: There are a lot TODOs in the official example. This makes it even harder for me to use the library, as it seems broken/unfinished on some parts?

The 1.0 release despite being the only stable release is fairly old (6 months). There were many improvements since then. Unfortunately because there were a fair amount of breaking changes too since then in the API, in particular for multi-track support, and we don't have enough resources to work on both 1.0 and master, the TestAppUWP in 1.0 still has some work left that we will probably never finish. I suggest looking also at the TestAppUWP on master; although it uses a different API, the underlying WebRTC concepts like signaling and negotiation didn't change, and for the most part the API around that is the same. I strongly suggest also looking at the WebRTC 1.0 standard and other WebRTC sources, which again share the same concepts and can help understanding how WebRTC works in general (which is a lot more complex than it may seem at first). See in particular the examples (section 10) and [this call flow diagram](https://www.w3.org/TR/webrtc/#call-flow-browser-to-browser).
